lib/ukfallocbuddy/tests/test_ukfallocbuddy.c

```diff
@@ -0,0 +1,179 @@
+	void *start_zone = malloc(10000);
+	__paddr_t start;
+	__sz len = 10000;
+	void *fa_meta = malloc(10000);
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+	int falloc_result;
+
+	falloc_result = fa->falloc(fa, (__paddr_t *)start, 1, 0);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_result, 0);
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
+}
+
+UK_TESTCASE(ukfallocbuddy, test_alloc_from_range)
+{
+
+	// test allocation after a falloc
+	void *start_zone = malloc(50000);
+	__paddr_t start, min, max;
+	__sz len = 50000;
+	void *fa_meta = malloc(10000);
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+	int falloc_from_range_result;
+
+	min = start + 4096;
+	max = start + 20000;
+	fa->falloc(fa, (__paddr_t *)start, 1, FALLOC_FLAG_ALIGNED);
+	falloc_from_range_result = fa->falloc_from_range(fa, (__paddr_t *)start, 2, 0, min, max);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, 0);
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
+}
+
+UK_TESTCASE(ukfallocbuddy, test_alloc_large)
+{
+
+	// test allocation with too many pages
+	void *start_zone = malloc(50000);
+	void *fa_meta = malloc(10000);
+	__paddr_t start;
+	__sz len = 50000;
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+	int falloc_result;
+
+	falloc_result = fa->falloc(fa, (__paddr_t *)start, 600, 0);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_result, -14);
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
+}
+
+UK_TESTCASE(ukfallocbuddy, test_alloc_over)
+{
+
+	// test allocation over an already allocated zone
+	void *start_zone = malloc(50000);
+	void *fa_meta = malloc(10000);
+	__paddr_t start;
+	__paddr_t end;
+	__sz len = 50000;
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+	int falloc_from_range_result;
+
+	end = start + 8192;
+	fa->falloc(fa, (__paddr_t *)start, 2, 0);
+	falloc_from_range_result = fa->falloc_from_range(fa, (__paddr_t *)start, 2, 0, start, end);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, -12);
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
+}
+
+UK_TESTCASE(ukfallocbuddy, test_alloc_over_2)
+{
+
+	// test allocation of two frames between an already allocated one
+	void *start_zone = malloc(50000);
+	void *fa_meta = malloc(10000);
+	__paddr_t start;
+	__sz len = 50000;
+	int falloc_from_range_result;
+	struct uk_falloc *fa = generate(fa_meta, start_zone, len, &start);
+
+	fa->falloc_from_range(fa, (__paddr_t *)start, 1, 0, start + 4097, start + 8192);
+	falloc_from_range_result = fa->falloc_from_range(fa, (__paddr_t *)start, 2, 0, start, start + 12288);
+	UK_TEST_EXPECT_SNUM_EQ(falloc_from_range_result, 0);
+	free(start_zone);
+	free(fa_meta);
+	free(fa);
+}
+uk_testsuite_register(ukfallocbuddy, NULL);
+
+struct uk_falloc *generate(void *fa_meta, void *start_zone, unsigned long len, __paddr_t *new_start)
+{
+	__sz fa_meta_size;
+	__vaddr_t dm_off;
+	__paddr_t start = (__paddr_t)start_zone;
+	unsigned long pages = len >> PAGE_SHIFT;
+	struct uk_falloc *fa = malloc(100000);
```

Comment:
See the comment above about allocating the required space using `uk_fallocbuddy_size()`.